### PR TITLE
Fix SBR unreachable observer cleanup

### PR DESCRIPTION
**Pull request by Arkatufus**
Fix edge case where cluster node become unreachable after gossiping about another unreachable node, causing the cluster leader to erroneously decide that the cluster is irreparably unstable and downs the whole cluster.

* Node-1 is leader
* Node-2 went unreachable to the whole cluster
* Node-3 gossips that Node-2 is unreachable
* Node-1 SBR records Node-3 as an observer
* Node-3 went unreachable to the whole cluster before SBR stable-after interval expires
* Node-1 SBR stable-after interval expires
* Node-1 thinks that Node-3 is indirectly connected to the cluster and couldn't resolve the graph
* Node-1 SBR downs all nodes

## Changes

Add a check in `DowningStrategy.AdditionalNodesToDownWhenIndirectlyConnected` to prune out records where its `Record.Observer` is proven to be unreachable by consensus with all other records that are known to be reachable.

## Log data

```
[01:07:26 WRN] SBR double Akka.Cluster.SBR.DownIndirectlyConnected decision, downing all instead. originalReachability: [Reachability([akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6009 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6003 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6003 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6003 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6011 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6011 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6011 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6013 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6013 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6013 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6004 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6004 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6004 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6001 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6001 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6001 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6014 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6014 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6014 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (2)])], filtered reachability [Reachability([akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6010 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6009 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6003 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6003 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6003 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6011 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6011 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6011 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6015 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6002 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (2)][akka.tcp://AkkaCluster@localhost:6013 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6013 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6013 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6004 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6004 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6004 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6001 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6001 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6001 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6014 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6014 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (2)], [akka.tcp://AkkaCluster@localhost:6014 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6006, 2095747650): Unreachable [Unreachable] (1)][akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932): Unreachable [Unreachable] (4)], [akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6008, 1606735766): Unreachable [Unreachable] (3)], [akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6005, 1615201291): Unreachable [Unreachable] (1)], [akka.tcp://AkkaCluster@localhost:6012 -> UniqueAddress: (akka.tcp://AkkaCluster@localhost:6007, 1615201291): Unreachable [Unreachable] (2)])], still indirectlyConnected: [UniqueAddress: (akka.tcp://AkkaCluster@localhost:6009, 1134654932)], seenBy: [akka.tcp://AkkaCluster@localhost:6011, akka.tcp://AkkaCluster@localhost:6014, akka.tcp://AkkaCluster@localhost:6010, akka.tcp://AkkaCluster@localhost:6012, akka.tcp://AkkaCluster@localhost:6003, akka.tcp://AkkaCluster@localhost:6015, akka.tcp://AkkaCluster@localhost:6002, akka.tcp://AkkaCluster@localhost:6001, akka.tcp://AkkaCluster@localhost:6013, akka.tcp://AkkaCluster@localhost:6004]
```

### CODE REVIEW

File: src/core/Akka.Cluster/SBR/DowningStrategy.cs [hunk 1]
```diff
@@ -229,10 +229,40 @@ private ImmutableHashSet<UniqueAddress> AdditionalNodesToDownWhenIndirectlyConne
...
+                    var records = originalReachability.Records
+                        .Where(r => localSeenBy.Contains(r.Observer.Address) && r.Subject.Equals(address))
+                        .Select(r => r.Observer)
+                        .ToImmutableHashSet();
+                    
+                    // Add the node to the prune list if we reach a consensus
+                    if(records.Count == localSeenBy.Count)
+                        pruneList.Add(address);
+                }
+                #endregion
```
Comment (Arkatufus) [inline src/core/Akka.Cluster/SBR/DowningStrategy.cs]: This algorithm collects record observer addresses that are known to be unreachable by consensus.

File: src/core/Akka.Cluster/SBR/DowningStrategy.cs [hunk 2]
```diff
@@ -229,10 +229,40 @@ private ImmutableHashSet<UniqueAddress> AdditionalNodesToDownWhenIndirectlyConne
...
+                
                 Reachability = Reachability.FilterRecords(
                     r =>
                         // we only retain records for addresses that are still downable
                         downable.Contains(r.Observer) && downable.Contains(r.Subject) &&
+                        // prune out records that are known to be disconnected islands
+                        !pruneList.Contains(r.Observer) &&
```
Comment (Arkatufus) [inline src/core/Akka.Cluster/SBR/DowningStrategy.cs:239]: Prune out records that Observer are known to be unreachable by consensus

**Review by Arkatufus — COMMENTED** [review]
Self review

File: src/core/Akka.Cluster/SBR/DowningStrategy.cs [hunk 3]
```diff
@@ -256,6 +260,51 @@ private ImmutableHashSet<UniqueAddress> AdditionalNodesToDownWhenIndirectlyConne
...
+        /// <param name="unreachable">Local copy of the unreachable addresses</param>
+        /// <returns>Hash set of detected disconnected addresses</returns>
+        private ImmutableHashSet<UniqueAddress> PruneUnreachableNodes(
+            Reachability reachability,
+            ImmutableHashSet<UniqueAddress> unreachable)
+        {
+            // Get all reachable nodes
+            var allReachable = AllMembers.Select(m => m.UniqueAddress)
+                .Where(a => !unreachable.Contains(a))
+                .ToImmutableHashSet();
```
Comment (Arkatufus) [inline src/core/Akka.Cluster/SBR/DowningStrategy.cs:280]: From the all cluster members set, exclude all members that are inside the unreachable set. This gives us all of the cluster members that is reachable.

File: src/core/Akka.Cluster/SBR/DowningStrategy.cs [hunk 4]
```diff
@@ -256,6 +260,51 @@ private ImmutableHashSet<UniqueAddress> AdditionalNodesToDownWhenIndirectlyConne
...
+                
+            // Collect all nodes that are reported as both record observer and unreachable 
+            // (possible indirect connection)
+            var possibleIndirect = reachability.Records
+                .Where(r => unreachable.Contains(r.Observer))
+                .Select(r => r.Observer)
+                .ToImmutableHashSet();
```
Comment (Arkatufus) [inline src/core/Akka.Cluster/SBR/DowningStrategy.cs:287]: From all the `Reachability` `Records` that we want to filter, find any `Record.Observer` that is also inside the unreachable set. These are all potential poison records that needs to be cleaned.

File: src/core/Akka.Cluster/SBR/DowningStrategy.cs [hunk 5]
```diff
@@ -256,6 +260,51 @@ private ImmutableHashSet<UniqueAddress> AdditionalNodesToDownWhenIndirectlyConne
...
+
+            // For each possible islands, reach a consensus with all nodes that are still reachable
+            // that they also could not see the possible island node.
+            var pruneList = new List<UniqueAddress>();
+            foreach (var address in possibleIndirect)
+            {
+                var records = reachability.Records
+                    .Where(r => allReachable.Contains(r.Observer) && r.Subject.Equals(address))
+                    .Select(r => r.Observer)
+                    .ToImmutableHashSet();
```
Comment (Arkatufus) [inline src/core/Akka.Cluster/SBR/DowningStrategy.cs:297]: Based on the `Reachability.Records` that we have, if all of the nodes inside the all reachable set agrees that they also could not reach the potential poison record, prune it from the `Reachability.Records`

**Review by Arkatufus — COMMENTED** [review]
Self Review

**Review by Aaronontheweb — APPROVED** [review]
LGTM - this should harden the SBR system against indirectly connected node cases where an observer managed to mark some gossip as seen on its way out the door.

**Pull request merged**